File: 0186-reverse-words-in-a-string-ii/0186-reverse-words-in-a-string-ii.cpp
class Solution {
public:
    void reverseWords(vector<char>& s) {
        reverse(s.begin(), s.end()); 
        int start = 0, end = 0; 
        while(end < s.size()){
            if(isalnum(s[end])){
                start = end; 
                while(end < s.size() && isalnum(s[end])){
                    end++; 
                }
                
                int tmp = end-1; 

                while(start < tmp){
                    swap(s[start], s[tmp]);
                    start++;
                    tmp--;  
                }
            }
            end++; 
        }
    }
};

File: 0186-reverse-words-in-a-string-ii/0186-reverse-words-in-a-string-ii.java
class Solution {
    public void reverse(char[] s, int start, int end){
        while(start < end){
            char tmp = s[start];
            s[start] = s[end];
            s[end] = tmp;
            start++;
            end--; 
        }
    }
    public void reverseWords(char[] s) {
        reverse(s,0,s.length-1); 
        int start = 0, end = 0; 
        while(end < s.length){
            if(!Character.isLetterOrDigit(s[end])){
                reverse(s, start,end-1); 
                start = end + 1; 
            }
            end++; 
        }
        reverse(s, start, end - 1); 
    }
}